test-suite/js/src/test/scala/org/scalajs/testsuite/jsinterop/JSSymbolTest.scala

```diff
@@ -0,0 +1,345 @@
+    assertEquals("SJSDefinedInnerObject.innerObject", obj0.innerObject.toString())
+
+    val obj1: InnerObjectTrait = new SJSDefinedInnerObject
+    assertEquals("object", js.typeOf(obj1.innerObject.asInstanceOf[js.Any]))
+    assertEquals("SJSDefinedInnerObject.innerObject", obj1.innerObject.toString())
+  }
+
+  @Test def native_with_methods(): Unit = {
+    val obj = mkObject(
+        sym1 -> ((x: Int) => x + 2),
+        sym2 -> ((x: String) => "Hello " + x)
+    )
+
+    assertEquals(4, obj.asInstanceOf[MethodClass].foo(2))
+    assertEquals("Hello World", obj.asInstanceOf[MethodClass].bar("World"))
+
+    assertEquals(6, obj.asInstanceOf[MethodTrait].foo(4))
+    assertEquals("Hello Moon", obj.asInstanceOf[MethodTrait].bar("Moon"))
+  }
+
+  @Test def sjsdefined_with_methods(): Unit = {
+    val obj = new SJSDefinedMethod
+    assertEquals(4, obj.foo(2))
+    assertEquals("Hello World", obj.bar("World"))
+
+    val asTrait: MethodTrait = obj
+    assertEquals(6, asTrait.foo(4))
+    assertEquals("Hello Moon", asTrait.bar("Moon"))
+  }
+
+  @Test def native_with_overloaded_methods(): Unit = {
+    val obj = mkObject(
+        sym1 -> ((x: Int) => x + 3),
+        sym2 -> ((x: String) => "Hello " + x)
+    )
+
+    assertEquals(5, obj.asInstanceOf[OverloadedMethodClass].foo(2))
+    assertEquals("Hello World", obj.asInstanceOf[OverloadedMethodClass].foo("World"))
+
+    assertEquals(6, obj.asInstanceOf[OverloadedMethodTrait].foo(3))
+    assertEquals("Hello Moon", obj.asInstanceOf[OverloadedMethodTrait].foo("Moon"))
+  }
+
+  @Test def sjsdefined_with_overloaded_methods(): Unit = {
+    val obj = new SJSDefinedOverloadedMethod
+    assertEquals(5, obj.foo(2))
+    assertEquals("Hello World", obj.foo("World"))
+
+    val asTrait: OverloadedMethodTrait = obj
+    assertEquals(6, asTrait.foo(3))
+    assertEquals("Hello Moon", asTrait.foo("Moon"))
+  }
+
+  @Test def native_with_overloaded_runtime_dispatch_methods(): Unit = {
+    val obj = mkObject(
+        sym1 -> { (x: Any) =>
+          x match {
+            case x: Int    => x + 3
+            case x: String => "Hello " + x
+          }
+        }
+    )
+
+    assertEquals(5,
+        obj.asInstanceOf[OverloadedRuntimeDispatchMethodClass].foo(2))
+    assertEquals("Hello World",
+        obj.asInstanceOf[OverloadedRuntimeDispatchMethodClass].foo("World"))
+
+    assertEquals(6,
+        obj.asInstanceOf[OverloadedRuntimeDispatchMethodTrait].foo(3))
+    assertEquals("Hello Moon",
+        obj.asInstanceOf[OverloadedRuntimeDispatchMethodTrait].foo("Moon"))
+  }
+
+  @Test def sjsdefined_with_overloaded_runtime_dispatch_methods(): Unit = {
+    val obj = new SJSDefinedOverloadedRuntimeDispatchMethod
+    assertEquals(5, obj.foo(2))
+    assertEquals("Hello World", obj.foo("World"))
+
+    val asTrait: OverloadedRuntimeDispatchMethodTrait = obj
+    assertEquals(6, asTrait.foo(3))
+    assertEquals("Hello Moon", asTrait.foo("Moon"))
+  }
```

Comment:
Hum, yes. It's kind of a consequence of the fact that reading via the trait works + the previous test passes. But it's not very self-contained. I'll add a "dynamic"-based test as well.